create function Elo rating

sanyamaks opened the pull request:
Сама функция не сложная, а вот над тем как получить данные пришлось подумать, скорее даже в организационном плане. Если все данные будут браться действительно с games.json со всеми играми, то, на мой взгляд, нужно ставить какой-то маяк `wasCounted` о том, был ли пересчитан рейтинг или нет.
Т.к. далее куда передавать данные, тоже не понятно, пока оставил так до кодревью.
Файл games.json с добавленным ключем rating залил в комментарии к задаче в Trello, т.к. здесь json не поддерживается.

sanyamaks committed:
create function Elo rating

akuznetsov-md commented:
Для хранения данных сделаем отдельную таблицу `UsersGameRating`. В каждой строке этой таблицы будет храниться ID игры, ID всех игроков, рейтинги всех игроков к моменту игры и количество заработанных баллов каждым. 

Это даст возможность оценивать силы игроков в прошедших играх и составлять график изменения рейтинга.

Текущий рейтинг игрока будем хранить отдельным полем в `Users`.

С т.з. кода API метод создания очередной игры будет выглядеть как-то так

```
.post("/api/game", async ctx => {
  // Добавляем очередную игру
  const game = await gamesModule.addGame(ctx.request.body);

  // Регистрируем игру в таблице учета рейтинга.
  // Внутри модуля ratingModule вычисляем очки, которые заработали игроки.
  // В результате добавляется запись в таблицу UsersGameRating, а также обновляется актуальное значение рейтинга каждого игрока
  const gameRating = await ratingModule.addGame(game);

  // Возвращаем объект созданной игры и объект с данными о перераспределении рейтинга
  ctx.body = { game, gameRating };
})
```

На данном этапе тебе нужно создать модуль `ratingModule`, который принимает игру, а возвращает объект в формате:

```
{
  gameId,
  userA1Id,
  userA2Id,
  userB1Id,
  userB2Id,
  userA1Rating,
  userA2Rating,
  userB1Rating,
  userB2Rating,
  userA1Points,
  userA2Points,
  userB1Points,
  userB2Points
}
```

Запись в БД реализуем в следующей итерации.

Также стоит отделить модуль от тестовых данных. Попробуй сделать отдельный API эндпоинт `POST /api/gameRating/test`, который принимает в body объект игры, а в ответе возвращает результат работы твоей функции.

sanyamaks commented:
Что будет храниться в `userA1Points` и остальных `...Points` ? Именно заработанные очки, а не получившийся рейтинг? Если так, то получается суммирование этих очков будет происходить где-то в другом месте?

akuznetsov-md commented:
Да, заработанные очки.

Суммирование и назначение нового рейтинга конкретному игроку будет внутри `ratingModule` — это пока можно опустить

sanyamaks committed:
Merge pull request #1 from kuznetsov0209/master

Update

sanyamaks committed:
Merge pull request #2 from sanyamaks/master

Update

sanyamaks committed:
Created ratingModule.js and apiGameRatingRouter.js

sanyamaks commented:
Создал запрос в postman для проверки, надеюсь, откроется(если нет, то объект игры с ретйтингом можно взять в комментарии к карточке в trello):  
https://blue-robot-5461.postman.co/collections/9178059-02d538d9-46ef-41fd-b422-07ddb1f2b80e?version=latest&workspace=01a3c26d-8a63-42e1-8ea5-d54ac21ad7c1

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,98 @@
+    let newRatingTeamA =
+      ratingTeamA + MAXIMUM_POINTS * (actualPoints - expectedPoints);
+
+    let deltaA = newRatingTeamA - ratingTeamA;
```
А зачем ты сначала прибавляешь ratingTeamA, а затем вычитаешь?

sanyamaks replied:
У меня нет объяснения :D 
Видимо, глаз замылидся, и такую глупую ошибку не заметил.
Исправлю

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,98 @@
+    let userA1Id = usersId[0],
+      userA2Id = usersId[1],
+      userB1Id = usersId[2],
+      userB2Id = usersId[3];
```
Я обычно использую `const` вместо `let`. Также я предпочитаю каждую переменную объявлять самостоятельно. У меня нет рационального объяснения, почему так лучше, и вообще это дело вкуса. Но реальность такова, что в проектах такие спорные вопросы обычно решаются диктатурой лида. Поэтому давай следовать моему вкусу :)

akuznetsov-md replied:
Кстати, в JS можно деструктурировать массивы:
```
const [userA1Id, userA2Id, userB1Id, userB2Id] = usersId;
```

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  function elo(gameId, usersId, usersRating, isVictoryTeamA) {
+    const MAXIMUM_POINTS = 16;
```
А оставь ссылку на источник этой цифры.

akuznetsov-md replied:
Ну или в целом на источник информации о рейтинге Эло, по которому ты составлял эти функции

sanyamaks replied:
Закомментировать в коде или кинуть сюда?

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,98 @@
+  );
+  return gameRating;
+
+  function findOutWinner(goals, usersId) {
```
У функции `findOutWinner` есть ряд проблем.

1. возвращает булевое значение: "найти победителя (голы, пользователь) => да" — странно читается
2. принимает странный набор аргументов — список голов и список всех игроков без разделения на команды — как понять, кто против кого играл и кто победил?

sanyamaks replied:
1.Тогда стоит ли назвать функцию isVictoryTeamA, например?

2.Подразумевается, что первые 2 игрока, первая команда, а вторые 2 - другая. Понимаю, что не очевидно. Стоит привязать их как-то к свойству "team" через условие?

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,98 @@
+    usersId,
+    usersRating,
+    findOutWinner(goals, usersId)
+  );
```
Метод для вычисление рейтинга по конкретному алгоритму должен быть максимально обособлен от приложения. gameId, usersId — это все специфично для конкретного приложения кикера. Сделай функцию так, чтобы ее можно было без проблем использовать для вычисления рейтинга при игре в шахматы и в совершенно другом формате хранения данных.

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,98 @@
+// const moment = require("moment");
+// const db = require("../models");
+
+async function countPoints(game) {
```
countPoints не лучшее название. По факту эта функция будет добавлять игру в систему учета рейтинга. "система учета рейтинга" у нас отражена в названии файла — ratingModule. Соответственно саму функцию следует назвать addGame или addItem

akuznetsov-md commented on services/elo.js:
```diff
@@ -0,0 +1,77 @@
+  // console.log({newRating1TeamA, newRating2TeamA, newRating1TeamB, newRating2TeamB});
+
+  return { newRating1TeamA, newRating2TeamA, newRating1TeamB, newRating2TeamB };
+}
```
Если этот файл уже не используется, то его следует удалить

akuznetsov-md replied:
Хотя на самом деле функцию `elo` все-таки следует вынести из ratingModule и хранить здесь — в сервисах. В какой-то момент мы можем захотеть использовать другую, более совершенную методику. В таком случае, мы просто сделаем еще один сервис и подключим его в ratingModule, не меняя формат функций и формат хранения данных.

sanyamaks committed:
Fixed elo.js and ratingModule.js

sanyamaks commented:
Исправил замечания. Посмотри, пожалуйста.

akuznetsov-md commented on app/ratingModule.js:
```diff
@@ -0,0 +1,95 @@
+      });
+    return scoredTeamA > scoredTeamB;
+  }
+}
```
Не знаю, как-то все равно сложновато читается функция. Попробуй такой вариант

```js
const elo = require("../services/elo");

const TEAM_A = 0;
const TEAM_B = 1;

async function addGame(game) {
  const usersByTeam = {
    [TEAM_A]: game.Users.filter(user => user.GamePlayer.team === TEAM_A),
    [TEAM_B]: game.Users.filter(user => user.GamePlayer.team === TEAM_B)
  };

  const usersIdsByTeam = {
    [TEAM_A]: usersByTeam[TEAM_A].map(user => user.id),
    [TEAM_B]: usersByTeam[TEAM_B].map(user => user.id)
  };

  const scoreByTeam = {
    [TEAM_A]: game.Goals.filter(
      goal =>
        (usersIdsByTeam[TEAM_A].includes(goal.UserId) && !ownGoal) ||
        (!usersIdsByTeam[TEAM_A].includes(goal.UserId) && ownGoal)
    ),
    [TEAM_B]: game.Goals.filter(
      goal =>
        (usersIdsByTeam[TEAM_B].includes(goal.UserId) && !ownGoal) ||
        (!usersIdsByTeam[TEAM_B].includes(goal.UserId) && ownGoal)
    )
  };

  const ratingByTeam = {
    [TEAM_A]: usersByTeam[TEAM_A][0].rating + usersByTeam[TEAM_A][1].rating,
    [TEAM_B]: usersByTeam[TEAM_B][0].rating + usersByTeam[TEAM_B][1].rating
  };

  // 1. Функция должна начинаться с глагола. Поэтому вместо elo.elo лучше написать elo.calculatePoints
  // 2. Elo — это универсальный способ посчитать очки рейтинга в игре 1 на 1. Поэтому имеет смысл использовать абстрактные названия player1 и player2
  // 3. При деструктуризации объекта можно переименовать названия переменных. В данном случае я явно указал, что очки посчитаны по формуле Elo
  // 4. Имеет смысл передавать счет, вместо булевого результата матча. Так у нас будет возможность корректировки формулы в зависимости от счета. Все-таки для 10:0 и 10:9 должно быть разное количество очков
  const {
    player1Points: eloPlayer1Points,
    player2Points: eloPlayer2Points
  } = elo.calculatePoints({
    player1: {
      rating: ratingByTeam[TEAM_A],
      score: scoreByTeam[TEAM_A]
    },
    player2: {
      rating: ratingByTeam[TEAM_B],
      score: scoreByTeam[TEAM_B]
    }
  });

  return {
    gameId: game.id,
    userA1Id: usersByTeam[TEAM_A][0].id,
    userA2Id: usersByTeam[TEAM_A][1].id,
    userB1Id: usersByTeam[TEAM_B][0].id,
    userB2Id: usersByTeam[TEAM_B][1].id,
    userA1Rating: usersByTeam[TEAM_A][0].rating,
    userA2Rating: usersByTeam[TEAM_A][1].rating,
    userB1Rating: usersByTeam[TEAM_B][0].rating,
    userB2Rating: usersByTeam[TEAM_B][1].rating,
    userA1Points: eloPlayer1Points / 2,
    userA2Points: eloPlayer1Points / 2,
    userB1Points: eloPlayer2Points / 2,
    userB2Points: eloPlayer1Points / 2
  };
}

module.exports = {
  addGame
};

```

sanyamaks replied:
А разве по вычислению рейтинга Эло есть разница, 10:0 или 10:9? Там же учитывается только выйгрыш, ничья или проигрыш.

akuznetsov-md replied:
Нам следует адаптировать эту формулу под кикер. Задача не в том, чтобы прикрутить Эло, а в том чтобы реализовать адекватную систему подсчета рейтинга

sanyamaks committed:
Fixed RatingModule.js and elo.js

akuznetsov-md commented on services/elo.js:
```diff
@@ -0,0 +1,26 @@
+  const delta = MAXIMUM_POINTS * (actualPoints - expectedPoints);
+
+  const player1Points = delta;
+  const player2Points = +-delta;
```
А что делает конструкция `+-`?

sanyamaks replied:
Я возможно сейчас ошибаюсь, но когда проверял самый первый раз, он мне возвращал строку, а не число, поэтому сделал поставил + для преобразования в число

sanyamaks replied:
Из-за toFixed в предыдущем коммите возвращалась строка , а не число. Сейчас же возвращаться должно у нас число, и я просто оставил так по невнимательности.

akuznetsov-md commented on services/elo.js:
```diff
@@ -0,0 +1,26 @@
+  const expectedPoints = (
+    1 /
+    (1 + 10 ** ((player2.rating - player1.rating) / 400))
+  ).toFixed(2);
```
`toFixed` приводит число в строку. Есть ли какой-то смысл в этом?

sanyamaks replied:
Выше исправил комментарий.
Касательно  + смысла нету. 
toFixed применял для округления.

akuznetsov-md replied:
> toFixed применял для округления

Это некорректный способ округления. `toFixed` нужен для форматирования числа при выводе. Для округления нужно использовать методы объекта `Math`.

akuznetsov-md commented on services/elo.js:
```diff
@@ -0,0 +1,26 @@
+
+  const actualPoints = player1.score > player2.score ? 1 : 0;
+
+  const expectedPoints = (
```
Имеет смысл указать, что мы рассчитываем рейтинг для игрока 1:

```
const player1ActualResult = ...
const player1ExpectedResult = ...

...
// отдельная переменная delta лишняя в данном случае

const player1Points = 
const player2Points = -player1Points

```

sanyamaks committed:
Fixed elo.js:
Deleted delta
Replaced actualPoints with player1ActualResult
Replaced expectedPoints with player1ExpectedResult
